### ISSUE TITLE
Add the plural annotation during model generation

### ISSUE THREAD

**Issue by shawkins** [fabric8io/kubernetes-client#3589]
At a minimum that plural annotation should be added based upon the existing pluralize logic.  It does not appear there is a good way to extract the plural from the go classes - the plural seems to only be referenced in the client.

_Originally posted by @manusa in https://github.com/fabric8io/kubernetes-client/issues/3539#issuecomment-967131909_

**Comment by manusa** [comment]
I recall I tried to check that information once, but couldn't find it.
However, it must be somewhere. The Swagger/OpenAPI generator **does** get this information from some place, and sometimes it doesn't even fall into any kind of pluralize logic. Maybe someone from the ~~dark~~Go side knows where. We can also try to allocate some time to figure this out.

**Comment by shawkins** [comment]
Tracing this back a few steps.  The docs are built from a swagger https://github.com/kubernetes-sigs/reference-docs/tree/master/gen-apidocs/config/v1_23

The swagger comes from the main build https://github.com/kubernetes/kubernetes/tree/v1.23.0/api

I think this file is updated from the server itself - https://github.com/kubernetes/kubernetes/blob/0d3f2ca371d03cc58c36d3ee3523b719e987fc35/hack/update-openapi-spec.sh 

So the plurals are then coming from the api-server openapi support 
https://github.com/kubernetes/kubernetes/blob/f784891854ccb8d904a3f6d5da1e8e4fc940dcb2/staging/src/k8s.io/apiserver/pkg/server/routes/openapi.go

Which builds that swagger off of the known WebServices.  I haven't tracked down where the WebServices are being built - but their path will of course contain the plural.  @manusa do you know who would be good to reach out to, or should I just keep digging?

**Comment by manusa** [comment]
> So the best we can do is somehow join the current logic with the processing of the swagger.json to parse out / resolve the plurals.

:sob: For me this is the worst possible outcome, since it will require both processing swagger and go.

The issue is even worst for some of the extensions. The swagger OpenAPI docs are incomplete and for them. I really can't understand how this might actually work for Tekton etc. (e.g. dealing with Tekton resources from kubectl, how does kubectl know where the endpoint is if the OpenAPI is incomplete?)

**Comment by shawkins** [comment]
> I really can't understand how this might actually work for Tekton etc.

Wouldn't tekton be installed via crds?  My understanding is that processing would install the endpoint handler for the plural and thus update the swagger.

**Comment by manusa** [comment]
The generated swagger docs are wrong/incomplete for Tekton.

I forgot about the APIResource endpoint(s). Kubectl was using this, so it's no problem (just like for us after the GenericResource enhancements).

But then, how is this endpoint getting the info for the plural, etc?

**Comment by shawkins** [comment]
> But then, how is this endpoint getting the info for the plural, etc?

For built-in types the plural is coming from the storage keys.

For crds, it's being built here https://github.com/kubernetes/kubernetes/blob/07d3a92ce6bc6d25f66a15a4f19294a729bcb90b/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/builder/builder.go#L142

**Comment by manusa** [comment]
So it all sums up to what you said before:

> That's probably the synopsis here - you need the default swagger + extension crds, or a running server with everything installed to determine the plurals. There is no built-in exposure of the plural information on just the models.

If we want to implement this, for the core APIs we can rely on the hardcoded swagger.json file.

For the extensions (+OpenShift) it would be much harder.

Is it still worth exploring this?
There were also a few KEPs/issues related to this (I don't have the references now). Maybe this can be blocked until a better alternative is implemented upstream.

**Comment by shawkins** [comment]
> Is it still worth exploring this?

Probably not.  The simplest thing we could do otherwise would be to add a validation as part of a test - for whatever resources are exposed by the openshift/minikube that's running, use logic similar to 
```
  GenericKubernetesResource resource = new GenericKubernetesResource();
  resource.setKind(kind);
  resource.setApiVersion(apiVersion);
  Object value = Serialization.unmarshal(Serialization.asJson(resource));
  Class<T> parsedType = (Class<T>) value.getClass();
  if (!parsedType.equals(GenericKubernetesResource.class)) {
    ResourceDefinitionContext rdc = ResourceDefinitionContext.fromResourceType(parsedType);
    // validate plural
  } // else doesn't have a class
```

**Comment by shawkins** [comment]
Marking as closed based upon #3664 - that's as much as we can do for now.
